### ISSUE TITLE
Setup fixed properties for heavily used properties

### ISSUE THREAD

**Issue by jthingelstad** [WikiApiary/WikiApiary#156]
Heavily used properties may benefit from having a fixed property assigned. First we need to figure out the properties that should be fixed.
- [ ] Has website
- [ ] Has interwiki URL
- [ ] Has interwiki prefix

**Comment by kghbln** [comment]
Usage is a factor and the ones with over 100,000 values are probably the first pick:

![image](https://cloud.githubusercontent.com/assets/1104078/3410976/3e1ac14e-fdf0-11e3-8167-6d3f15a8e831.png)

Thus there would be 18 since the special property  No. 2, 6 and 19 are already in a table of their own. This could be the first lot.

If we chose + 50,000 there would be 20 fixed properties in total.

If we chose + 25,000 there would be 28 fixed properties in total.

I does not really make sense to have fixed properties for 25,000 and lower I guess.

**Comment by plexish** [comment]
Looking over Special:Properties, there's quite a few between 15000 and 25000, mostly the one per website/one per active website properties so that'll grow significantly once big farms start getting mass-imported. It's probably worth looking very carefully at performance before and after making a bunch of the most used properties fixed to see whether it's likely to be worth fixing the lower use ones.

**Comment by jthingelstad** [comment]
@kghbln Has subobject should be a fixed property out of the box, would you agree? Seems like that should be a feature addition for Semantic MediaWiki.

**Comment by kghbln** [comment]
@thingles I just included this to show the top properties, however I marked this one as special property (see above) which is indeed using it's own table, at least I believe so. Same for No. 6 and 19.

@etesp I once talked to Nishay or Markus about fixed properties and we somehow agreed that 25.000 is a got usage count to start thinking of transforming it into a fixed property. If other properties move into this usage count this definitively has to be revisited. (Off topic: Still do not like to have Wikia on board - I see no information gain here. It is well known what they are doing)

**Comment by jthingelstad** [comment]
The 25,000 number would also be impacted by the performance characteristics of your DB server. A super-fast server with large memory and SSD would have a different threshold than a bargain-basement VPS with spinning disks and low memory.

**Comment by jthingelstad** [comment]
It’s amazing how much the Interwikimap data massively expanded the property count. Those ones over 1M are definitely first place to start.

**Comment by jthingelstad** [comment]
@kghbln how did you get the table you are showing there?

**Comment by kghbln** [comment]
> The 25,000 number would also be impacted by the performance characteristics of your DB server. A super-fast server with large memory and SSD would have a different threshold than a bargain-basement VPS with spinning disks and low memory.

Doesn't SSD allow for faster writes and seaches. If this is the case + 25.000 may very well be moved up

> how did you get the table you are showing there?

I just copied the contents of "Special:Properties" into a file, saved this one as csv and imported it into LibreOffice. The rest was just replacing text and sorting rows.

**Comment by jthingelstad** [comment]
That table should be a feature add to Semantic MediaWiki. It is a super helpful view, and would be even better with a column indicating fixed/dynamic property.

**Comment by kghbln** [comment]
Indeed, it would make things easier to be able to sort the properties by their usage count and to have an indication if we are taking about a fixed property (class fixed?) as it is already the case for special properties (class buildin)

**Comment by kghbln** [comment]
FYI @nischayn22

Edit: The first wiki I know of which will be using it.

**Comment by mwjames** [comment]
> Has subobject should be a fixed property out of the box, would you agree

`Has subobject` namely `_SOBJ` (table `smw_fpt_sobj`) is already a fixed property.

> That table should be a feature add to Semantic MediaWiki. It is a super helpful view, and would be even better with a column indicating fixed/dynamic property.

Looking for a developer who has time for implementing this as it doesn't require deep SMW/Store knowledge!

**Comment by kghbln** [comment]
Thanks for confirming that "Has subobject" is using it's own table. I would have been surprised if this were not the case.

**Comment by mwjames** [comment]
@kghbln @thingles This would be a good time to measure the impact of such change (whether we can actually see a significant read performance increase for those newly created fixed properties.)

**Comment by kghbln** [comment]
:+1: I believe we all are very curious to see the impact.

**Comment by nischayn22** [comment]
@kghbln I once was doing some work to have properties sorted by their counts. Unfortunately, never found time to complete that :(
https://gerrit.wikimedia.org/r/#/c/69306/

**Comment by kghbln** [comment]
@nischayn22 I did not know this. If it is possible for you tackling this again it will very much apprechiated. However, I added you to the loop since you were involved in adding the fixed property feature to SMW so you will proabably be interested in seeing how things go at WikiAPIary. I am excited about this.

**Comment by nischayn22** [comment]
@kghbln I just created an issue so that the discussion can continue there https://github.com/SemanticMediaWiki/SemanticMediaWiki/issues/357
I am not sure if I will take on this task as I want to first complete another task on SMW I have made my hands dirty with.

**Comment by jthingelstad** [comment]
@mwjames is there a recommended way to attempt to determine the performance improvement from the fixed properties? Can I use the self-reported times that Semantic MediaWiki puts in the query subobjects? I'd love to do a writeup on the performance gain but I’m not sure how to go about it.
